{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "8f7c80f9-10c3-4113-8366-6eccd29af6d8",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pickle\n",
    "import pandas as pd\n",
    "import multiprocessing as mp\n",
    "import numpy as np\n",
    "import pickle\n",
    "import matplotlib.pyplot as plt\n",
    "from Bio import SeqIO\n",
    "import os\n",
    "import subprocess\n",
    "import ast\n",
    "import warnings\n",
    "from pathlib import Path\n",
    "warnings.filterwarnings('ignore')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e22d07a9-2568-4cd8-a89f-151245bf112a",
   "metadata": {},
   "source": [
    "## Add hmmsearch hits of XP_021699539.1 vs XP_029735553.1\n",
    "XP_029735553.1 (Aedes albopictus) was a secondary chimera of XP_021699539.1 (Aedes aegypti) in the first pipeline iteration run on A. albopictus annotation release GCF_006496715.1, but was later marked as a lncRNA in the current A.albopictus annotation release. We confirmed its expression and sequence via RT-PCR and Sanger sequencing after the first iteration, so manually added XP_029735553.1 back for consideration as a secondary chimera of XP_021699539.1 in this iteration."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0c159db6-bc83-4e10-8bf2-547b9f951ba9",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[33mWARNING:\u001b[0m Skipping mount /var/lib/singularity/mnt/session/etc/resolv.conf [files]: /etc/resolv.conf doesn't exist in container\n",
      "# hmmsearch :: search profile(s) against a sequence database\n",
      "# HMMER 3.3.2 (Nov 2020); http://hmmer.org/\n",
      "# Copyright (C) 2020 Howard Hughes Medical Institute.\n",
      "# Freely distributed under the BSD open source license.\n",
      "# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -\n",
      "# query HMM file:                  XP_021699539.1_hmms\n",
      "# target sequence database:        outputs/suppressed_aedes_albopictus.fa\n",
      "# per-dom hits tabular output:     XP_021699539.1_hmms.domtblout\n",
      "# show alignments in output:       no\n",
      "# sequence reporting threshold:    E-value <= 0.01\n",
      "# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -\n",
      "\n",
      "Query:       GCF_002204515.2;XP_021699539.1;HGT_(5,150)  [M=129]\n",
      "Scores for complete sequences (score includes all domains):\n",
      "   --- full sequence ---   --- best 1 domain ---    -#dom-\n",
      "    E-value  score  bias    E-value  score  bias    exp  N  Sequence                       Description\n",
      "    ------- ------ -----    ------- ------ -----   ---- --  --------                       -----------\n",
      "    6.1e-99  313.6  21.3    2.6e-44  137.0   0.2    4.2  4  GCF_006496715.1;XP_029735553.1  uncharacterized protein LOC10941\n",
      "\n",
      "\n",
      "Domain annotation for each sequence:\n",
      ">> GCF_006496715.1;XP_029735553.1  uncharacterized protein LOC109414258 [Aedes albopictus]\n",
      "   #    score  bias  c-Evalue  i-Evalue hmmfrom  hmm to    alifrom  ali to    envfrom  env to     acc\n",
      " ---   ------ ----- --------- --------- ------- -------    ------- -------    ------- -------    ----\n",
      "   1 !  137.0   0.2   2.6e-44   2.6e-44       1     128 [.      12     142 ..      12     143 .. 0.97\n",
      "   2 !   96.6   0.5   7.8e-32   7.8e-32       1     126 [.     183     307 ..     183     310 .. 0.95\n",
      "   3 !   93.5   2.0   7.4e-31   7.4e-31       1     125 [.     377     502 ..     377     506 .. 0.95\n",
      "   4 ?   -3.9   0.8      0.96      0.96      97     125 ..    1047    1076 ..    1044    1080 .. 0.67\n",
      "\n",
      "\n",
      "\n",
      "Internal pipeline statistics summary:\n",
      "-------------------------------------\n",
      "Query model(s):                            1  (129 nodes)\n",
      "Target sequences:                          1  (1181 residues searched)\n",
      "Passed MSV filter:                         1  (1); expected 0.0 (0.02)\n",
      "Passed bias filter:                        1  (1); expected 0.0 (0.02)\n",
      "Passed Vit filter:                         1  (1); expected 0.0 (0.001)\n",
      "Passed Fwd filter:                         1  (1); expected 0.0 (1e-05)\n",
      "Initial search space (Z):                  1  [actual number of targets]\n",
      "Domain search space  (domZ):               1  [number of targets reported over threshold]\n",
      "# CPU time: 0.00u 0.00s 00:00:00.00 Elapsed: 00:00:00.00\n",
      "# Mc/sec: 72.29\n",
      "//\n",
      "Query:       GCF_002204515.2;XP_021699539.1;HGT_(190,334)  [M=139]\n",
      "Scores for complete sequences (score includes all domains):\n",
      "   --- full sequence ---   --- best 1 domain ---    -#dom-\n",
      "    E-value  score  bias    E-value  score  bias    exp  N  Sequence                       Description\n",
      "    ------- ------ -----    ------- ------ -----   ---- --  --------                       -----------\n",
      "   1.8e-133  425.9  16.9    7.2e-54  168.1   0.8    3.6  3  GCF_006496715.1;XP_029735553.1  uncharacterized protein LOC10941\n",
      "\n",
      "\n",
      "Domain annotation for each sequence:\n",
      ">> GCF_006496715.1;XP_029735553.1  uncharacterized protein LOC109414258 [Aedes albopictus]\n",
      "   #    score  bias  c-Evalue  i-Evalue hmmfrom  hmm to    alifrom  ali to    envfrom  env to     acc\n",
      " ---   ------ ----- --------- --------- ------- -------    ------- -------    ------- -------    ----\n",
      "   1 !  168.1   0.8   7.2e-54   7.2e-54       3     136 ..       6     143 ..       4     147 .. 0.95\n",
      "   2 !  167.1   2.1   1.5e-53   1.5e-53       1     136 [.     174     310 ..     174     317 .. 0.98\n",
      "   3 !   99.4   0.8   1.1e-32   1.1e-32       7     131 ..     375     501 ..     369     509 .. 0.88\n",
      "\n",
      "\n",
      "\n",
      "Internal pipeline statistics summary:\n",
      "-------------------------------------\n",
      "Query model(s):                            1  (139 nodes)\n",
      "Target sequences:                          1  (1181 residues searched)\n",
      "Passed MSV filter:                         1  (1); expected 0.0 (0.02)\n",
      "Passed bias filter:                        1  (1); expected 0.0 (0.02)\n",
      "Passed Vit filter:                         1  (1); expected 0.0 (0.001)\n",
      "Passed Fwd filter:                         1  (1); expected 0.0 (1e-05)\n",
      "Initial search space (Z):                  1  [actual number of targets]\n",
      "Domain search space  (domZ):               1  [number of targets reported over threshold]\n",
      "# CPU time: 0.00u 0.00s 00:00:00.00 Elapsed: 00:00:00.00\n",
      "# Mc/sec: 80.01\n",
      "//\n",
      "Query:       GCF_002204515.2;XP_021699539.1;HGT_(385,523)  [M=140]\n",
      "Scores for complete sequences (score includes all domains):\n",
      "   --- full sequence ---   --- best 1 domain ---    -#dom-\n",
      "    E-value  score  bias    E-value  score  bias    exp  N  Sequence                       Description\n",
      "    ------- ------ -----    ------- ------ -----   ---- --  --------                       -----------\n",
      "   1.1e-121  387.5  16.0    5.7e-49  151.9   0.3    3.7  3  GCF_006496715.1;XP_029735553.1  uncharacterized protein LOC10941\n",
      "\n",
      "\n",
      "Domain annotation for each sequence:\n",
      ">> GCF_006496715.1;XP_029735553.1  uncharacterized protein LOC109414258 [Aedes albopictus]\n",
      "   #    score  bias  c-Evalue  i-Evalue hmmfrom  hmm to    alifrom  ali to    envfrom  env to     acc\n",
      " ---   ------ ----- --------- --------- ------- -------    ------- -------    ------- -------    ----\n",
      "   1 !  151.9   0.3   5.7e-49   5.7e-49       2     130 ..       9     143 ..       8     152 .. 0.94\n",
      "   2 !  106.8   1.2   4.5e-35   4.5e-35       2     128 ..     180     308 ..     179     320 .. 0.93\n",
      "   3 !  136.6   2.3   2.9e-44   2.9e-44       5     136 ..     377     512 ..     374     517 .. 0.93\n",
      "\n",
      "\n",
      "\n",
      "Internal pipeline statistics summary:\n",
      "-------------------------------------\n",
      "Query model(s):                            1  (140 nodes)\n",
      "Target sequences:                          1  (1181 residues searched)\n",
      "Passed MSV filter:                         1  (1); expected 0.0 (0.02)\n",
      "Passed bias filter:                        1  (1); expected 0.0 (0.02)\n",
      "Passed Vit filter:                         1  (1); expected 0.0 (0.001)\n",
      "Passed Fwd filter:                         1  (1); expected 0.0 (1e-05)\n",
      "Initial search space (Z):                  1  [actual number of targets]\n",
      "Domain search space  (domZ):               1  [number of targets reported over threshold]\n",
      "# CPU time: 0.00u 0.00s 00:00:00.00 Elapsed: 00:00:00.00\n",
      "# Mc/sec: 87.43\n",
      "//\n",
      "Query:       GCF_002204515.2;XP_021699539.1;Meta_(980,1031)  [M=63]\n",
      "Scores for complete sequences (score includes all domains):\n",
      "   --- full sequence ---   --- best 1 domain ---    -#dom-\n",
      "    E-value  score  bias    E-value  score  bias    exp  N  Sequence                       Description\n",
      "    ------- ------ -----    ------- ------ -----   ---- --  --------                       -----------\n",
      "    4.2e-42  128.6  35.6    1.7e-34  104.3  12.2    5.1  4  GCF_006496715.1;XP_029735553.1  uncharacterized protein LOC10941\n",
      "\n",
      "\n",
      "Domain annotation for each sequence:\n",
      ">> GCF_006496715.1;XP_029735553.1  uncharacterized protein LOC109414258 [Aedes albopictus]\n",
      "   #    score  bias  c-Evalue  i-Evalue hmmfrom  hmm to    alifrom  ali to    envfrom  env to     acc\n",
      " ---   ------ ----- --------- --------- ------- -------    ------- -------    ------- -------    ----\n",
      "   1 !  104.3  12.2   1.7e-34   1.7e-34       1      63 []     979    1041 ..     979    1041 .. 0.99\n",
      "   2 !   14.7   2.0   1.5e-06   1.5e-06      16      45 ..    1049    1079 ..    1045    1084 .. 0.86\n",
      "   3 !   13.0   3.2   5.2e-06   5.2e-06      25      61 ..    1088    1125 ..    1082    1126 .. 0.81\n",
      "   4 !    8.0   0.1   0.00018   0.00018      14      45 ..    1137    1170 ..    1132    1176 .. 0.80\n",
      "\n",
      "\n",
      "\n",
      "Internal pipeline statistics summary:\n",
      "-------------------------------------\n",
      "Query model(s):                            1  (63 nodes)\n",
      "Target sequences:                          1  (1181 residues searched)\n",
      "Passed MSV filter:                         1  (1); expected 0.0 (0.02)\n",
      "Passed bias filter:                        1  (1); expected 0.0 (0.02)\n",
      "Passed Vit filter:                         1  (1); expected 0.0 (0.001)\n",
      "Passed Fwd filter:                         1  (1); expected 0.0 (1e-05)\n",
      "Initial search space (Z):                  1  [actual number of targets]\n",
      "Domain search space  (domZ):               1  [number of targets reported over threshold]\n",
      "# CPU time: 0.01u 0.00s 00:00:00.01 Elapsed: 00:00:00.00\n",
      "# Mc/sec: 37.42\n",
      "//\n",
      "[ok]\n"
     ]
    }
   ],
   "source": [
    "##validate that XP_029735553.1  is a hmmsearch hit to the metazoan and hgt intervals of XP_021699539.1 \n",
    "## XP_021699539.1_hmms is a concatenation of the hmms of XP_021699539.1  intervals found in outputs/hmmbuild\n",
    "##suppressed_aedes_albopictus.fa is a fasta of XP_029735553.1\n",
    "!singularity exec /cvmfs/singularity.galaxyproject.org/h/m/hmmer:3.3.2--he1b5a44_0 hmmsearch -E 1e-2 --cpu 2 --noali --domtblout \"XP_021699539.1_hmms.domtblout\" \"XP_021699539.1_hmms\"  \"outputs/suppressed_aedes_albopictus.fa\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "id": "8cecb246-4b0e-4e9e-ab5e-ef8bad92bd35",
   "metadata": {},
   "outputs": [],
   "source": [
    "!mkdir missing_aedes_albopictus_hmmsearch\n",
    "!sh scripts/split_hmmer_csv.sh \"XP_021699539.1_hmms.domtblout\" missing_aedes_albopictus_hmmsearch\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "id": "303ff03e-3c34-43ec-88b4-45e7c344892b",
   "metadata": {},
   "outputs": [],
   "source": [
    "!sh scripts/split_hmmer_csv.sh \"XP_021699539.1_hmms.domtblout\" missing_aedes_albopictus_hmmsearch"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 149,
   "id": "e0e1b267-9f33-4029-b692-3a83020df297",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "151.9\n",
      "104.3\n",
      "168.1\n",
      "96.6\n"
     ]
    }
   ],
   "source": [
    "##verify that all intervals of the query chimera GCF_002204515.2;XP_021699539.1 are significant hits to aedes albopictus GCF_006496715.1;XP_029735553.1\n",
    "for x in os.listdir('missing_aedes_albopictus_hmmsearch'):\n",
    "    if '.tsv' in x:\n",
    "        # original raw files: no header\n",
    "        arth = pd.read_csv(f'missing_aedes_albopictus_hmmsearch/{x}', sep=\"\\t\", header=None)\n",
    "        # drop unused cols\n",
    "        arth = arth.drop([1,4], axis=1)\n",
    "              # assign column names\n",
    "        arth.columns = cols\n",
    "        print(arth.domain_score.max())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "2832f476-9441-4efe-9c1e-7b6b3ed8a65e",
   "metadata": {},
   "outputs": [],
   "source": [
    "##load a dataframe of genome taxids from genome accessions\n",
    "df1=pd.read_csv('Data/genbank_genomes_4_22_2025.tsv',sep='\\t')\n",
    "df2=pd.read_csv('Data/refseq_genomes_scaffold_plus_4_19_2025.tsv',sep='\\t')\n",
    "dftax=pd.concat([df1,df2]).set_index('Assembly Accession')\n",
    "dftax.loc['GCF_006496715.1',['Organism Name','Organism Taxonomic ID']]=['Aedes albopictus',7160]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 99,
   "id": "b86a5e24-2194-43a9-89f8-9147a2f9e9df",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "##add hits to aedes albopictus GCF_006496715.1;XP_029735553.1 to main hmmsearch outputs for GCF_002204515.2;XP_021699539.1\n",
    "for x in os.listdir('missing_aedes_albopictus_hmmsearch'):\n",
    "    if '.tsv' in x:\n",
    "        # original raw files: no header\n",
    "        arth = pd.read_csv(f'missing_aedes_albopictus_hmmsearch/{x}', sep=\"\\t\", header=None)\n",
    "        # drop unused cols\n",
    "        arth = arth.drop([1,4], axis=1)\n",
    "        # assign column names\n",
    "        arth.columns = cols\n",
    "        # annotate arthropod hits with species & taxid\n",
    "        arth['species'] = [ dftax.loc[name.split(\";\")[0], 'Organism Name'] for name in arth['target_name'] ]\n",
    "        arth['taxid']   = [ dftax.loc[name.split(\";\")[0], 'Organism Taxonomic ID'] for name in arth['target_name'] ]\n",
    "\n",
    "        arth = arth.drop('description of target', axis=1)\n",
    "        ##drop e-values since database size differs\n",
    "        arth.loc[:,['E-value','c-Evalue', 'i-Evalue']]=np.nan\n",
    "        arthog=pd.read_csv(f'outputs/hmmsearch_v_arthropod/{x}',sep='\\t',index_col=0)\n",
    "        arth=arth.set_index('target_name')\n",
    "        pd.concat([arth,arthog]).sort_values('overall_score',ascending=False).to_csv(f'outputs/hmmsearch_v_arthropod/{x}',sep='\\t')\n",
    "        "
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python [conda env:.conda-rishabh]",
   "language": "python",
   "name": "conda-env-.conda-rishabh-py"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
