{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "d9b39eaa-99de-432e-923d-d845b4afdad6",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import os \n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9c8f82fa-de13-4d52-bebd-fdf9de558e92",
   "metadata": {},
   "source": [
    "## Map sequencing results (accessions) from pipeline iteration 1 to 2 "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "dcfe91f7-9ba2-4466-8227-078962099aa2",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_all=pd.read_csv(\"SI tables/SI table 2.tsv\",sep=\"\\t\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "6efa19c6-2a5e-41f3-a8df-5d8d56882566",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_all_origin=pd.read_csv(\"og_all_protein_tax_info.csv\",index_col=0)\n",
    "dfo=df_all_origin[df_all_origin.representative_seq=='TRUE']\n",
    "common=list(dfo[dfo.index.isin(df_all[df_all.representative].protein)].index)\n",
    "new_old={df_all[df_all.protein==x].cluster.values[0]:dfo.loc[x,'cluster'] for x in common}\n",
    "new_old=pd.DataFrame.from_dict(new_old,orient='index')\n",
    "new_old.index.name='new'\n",
    "new_old.columns=['old']\n",
    "new_old.to_csv('new_old_mapping.tsv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "21ddcf7e-ec45-40d6-b6fb-2f94dd187183",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "44"
      ]
     },
     "execution_count": 27,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "seq_og=pd.read_csv(\"SI tables/og_SI Table 11 - sequencing_results.tsv\",sep=\"\\t\")\n",
    "\n",
    "seq_og['reference protein accession ']=seq_og['reference protein accession '].replace({'KAH9406650':'KAH9406650.1','XP_019558301.2':' XP_019540707.3','XP_053634600.1':'XP_053634600.2','XP_042908388.1':'XP_015919340.2','XP_027231698.1':'XP_069990332.1','XP_027232647.1':'XP_027232647.2'})\n",
    "td=seq_og[seq_og['reference protein accession '].isin(df_all.protein)]\n",
    "len(td)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ba01ae74-4fdd-4c99-98c5-d5c07ff82550",
   "metadata": {},
   "source": [
    "## Make multiple sequence alignments"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "bbc40af1-1a37-4f86-b095-8e682f5e9682",
   "metadata": {},
   "outputs": [],
   "source": [
    "from Bio import Entrez, SeqIO\n",
    "\n",
    "def append_protein_to_fasta(accession, fasta_file, email='rkapoor@g.harvard.edu'):\n",
    "\n",
    "    Entrez.email = email\n",
    "    \n",
    "    # Fetch the protein sequence from the protein database in FASTA format\n",
    "    with Entrez.efetch(db=\"protein\", id=accession, rettype=\"fasta\", retmode=\"text\") as handle:\n",
    "        record = SeqIO.read(handle, \"fasta\")\n",
    "    \n",
    "    # Append the fetched record to the specified FASTA file\n",
    "    with open(fasta_file, \"a\") as output_handle:\n",
    "        SeqIO.write(record, output_handle, \"fasta\")\n",
    "    \n",
    "    print(f\"Protein {accession} has been appended to {fasta_file}.\")\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "f4b5aaf4-9866-4475-b66e-76bad81654d3",
   "metadata": {},
   "outputs": [],
   "source": [
    "from Bio import SeqIO"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "id": "7ef1463b-8636-403a-b34b-f15a3765b837",
   "metadata": {},
   "outputs": [],
   "source": [
    "##load amino acid translations from the genbank submission\n",
    "input_file = open(f\"genbank_translations_may_2025.fa\")\n",
    "fasta_dict = SeqIO.to_dict(SeqIO.parse(input_file, \"fasta\"))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "id": "c1aaf6f9-3b51-4cc2-90df-0d7495aecbb3",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Protein XP_035708168.1 has been appended to PCR_result_alignments/XP_035708168.1/protein.fa.\n",
      "Protein XP_046453153.1 has been appended to PCR_result_alignments/XP_046453153.1/protein.fa.\n",
      "Protein XP_045027829.1 has been appended to PCR_result_alignments/XP_045027829.1/protein.fa.\n",
      "Protein XP_046403459.1 has been appended to PCR_result_alignments/XP_046403459.1/protein.fa.\n",
      "Protein XP_050513096.1 has been appended to PCR_result_alignments/XP_050513096.1/protein.fa.\n",
      "Protein XP_042220148.1 has been appended to PCR_result_alignments/XP_042220148.1/protein.fa.\n",
      "Protein XP_029735553.1 has been appended to PCR_result_alignments/XP_029735553.1/protein.fa.\n",
      "Protein XP_046456339.1 has been appended to PCR_result_alignments/XP_046456339.1/protein.fa.\n",
      "Protein XP_027232647.2 has been appended to PCR_result_alignments/XP_027232647.2/protein.fa.\n",
      "Protein XP_023332299.1 has been appended to PCR_result_alignments/XP_023332299.1/protein.fa.\n",
      "Protein XP_037026007.1 has been appended to PCR_result_alignments/XP_037026007.1/protein.fa.\n",
      "Protein XP_023346081.1 has been appended to PCR_result_alignments/XP_023346081.1/protein.fa.\n",
      "Protein XP_037041958.1 has been appended to PCR_result_alignments/XP_037041958.1/protein.fa.\n",
      "Protein XP_037049533.1 has been appended to PCR_result_alignments/XP_037049533.1/protein.fa.\n",
      "Protein XP_037030969.1 has been appended to PCR_result_alignments/XP_037030969.1/protein.fa.\n",
      "Protein XP_035715507.1 has been appended to PCR_result_alignments/XP_035715507.1/protein.fa.\n",
      "Protein XP_023328891.1 has been appended to PCR_result_alignments/XP_023328891.1/protein.fa.\n",
      "Protein XP_046646423.1 has been appended to PCR_result_alignments/XP_046646423.1/protein.fa.\n",
      "Protein XP_037051404.1 has been appended to PCR_result_alignments/XP_037051404.1/protein.fa.\n",
      "Protein XP_011211954.1 has been appended to PCR_result_alignments/XP_011211954.1/protein.fa.\n",
      "Protein XP_053634600.2 has been appended to PCR_result_alignments/XP_053634600.2/protein.fa.\n",
      "Protein XP_032791327.2 has been appended to PCR_result_alignments/XP_032791327.2/protein.fa.\n",
      "Protein XP_046445252.1 has been appended to PCR_result_alignments/XP_046445252.1/protein.fa.\n",
      "Protein KAH9406650.1 has been appended to PCR_result_alignments/KAH9406650.1/protein.fa.\n",
      "Protein XP_023324156.1 has been appended to PCR_result_alignments/XP_023324156.1/protein.fa.\n",
      "Protein XP_001656415.1 has been appended to PCR_result_alignments/XP_001656415.1/protein.fa.\n",
      "Protein XP_046649021.1 has been appended to PCR_result_alignments/XP_046649021.1/protein.fa.\n",
      "Protein XP_046402901.1 has been appended to PCR_result_alignments/XP_046402901.1/protein.fa.\n",
      "Protein XP_021699539.1 has been appended to PCR_result_alignments/XP_021699539.1/protein.fa.\n",
      "Protein XP_037038002.1 has been appended to PCR_result_alignments/XP_037038002.1/protein.fa.\n",
      "Protein XP_069990332.1 has been appended to PCR_result_alignments/XP_069990332.1/protein.fa.\n",
      "Protein XP_023329593.1 has been appended to PCR_result_alignments/XP_023329593.1/protein.fa.\n",
      "Protein XP_029821973.3 has been appended to PCR_result_alignments/XP_029821973.3/protein.fa.\n",
      "Protein XP_015919340.2 has been appended to PCR_result_alignments/XP_015919340.2/protein.fa.\n"
     ]
    }
   ],
   "source": [
    "!mkdir 'PCR_result_alignments'\n",
    "td=seq_og[seq_og['reference protein accession '].isin(df_all.protein)]\n",
    "td=td[td['sequence name '].isin(fasta_dict.keys())]\n",
    "for a in set(td['reference protein accession ']):\n",
    "    !mkdir 'PCR_result_alignments'/\"$a\"\n",
    "    f=open(f'PCR_result_alignments/{a}/protein.fa','a')\n",
    "    for x in list(td[td['reference protein accession ']==a]['sequence name ']):\n",
    "        f.write(f\">{x}\\n\")\n",
    "        s=str(fasta_dict[x].seq)\n",
    "        f.write(f\"{s}\\n\")\n",
    "    f.close()\n",
    "    \n",
    "    append_protein_to_fasta(a,f'PCR_result_alignments/{a}/protein.fa')\n",
    "\n",
    "  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "id": "ff452a09-5432-43f9-b098-026cabc73154",
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    },
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "muscle 5.1.linux64 []  1056Gb RAM, 112 cores\n",
      "Built May 13 2023 06:17:56\n",
      "(C) Copyright 2004-2021 Robert C. Edgar.\n",
      "https://drive5.com\n",
      "\n",
      "Input: 4 seqs, avg length 296, max 394\n",
      "\n",
      "\n",
      "WARNING: Max OMP threads 1\n",
      "\n",
      "00:00 7.7Mb  CPU has 112 cores, running 1 threads\n",
      "00:00 11Mb    100.0% Calc posteriors\n",
      "00:00 12Mb    100.0% Consistency (1/2)\n",
      "00:00 12Mb    100.0% Consistency (2/2)\n",
      "00:00 12Mb    100.0% UPGMA5           \n",
      "00:00 12Mb    100.0% Refining\n",
      "\n",
      "muscle 5.1.linux64 []  1056Gb RAM, 112 cores\n",
      "Built May 13 2023 06:17:56\n",
      "(C) Copyright 2004-2021 Robert C. Edgar.\n",
      "https://drive5.com\n",
      "\n",
      "Input: 2 seqs, avg length 688, max 869\n",
      "\n",
      "\n",
      "WARNING: Max OMP threads 1\n",
      "\n",
      "00:00 7.7Mb  CPU has 112 cores, running 1 threads\n",
      "00:00 7.7Mb   100.0% Calc posteriors\n",
      "00:00 7.7Mb   100.0% UPGMA5         \n",
      "\n",
      "muscle 5.1.linux64 []  1056Gb RAM, 112 cores\n",
      "Built May 13 2023 06:17:56\n",
      "(C) Copyright 2004-2021 Robert C. Edgar.\n",
      "https://drive5.com\n",
      "\n",
      "Input: 3 seqs, avg length 1719, max 2477\n",
      "\n",
      "\n",
      "WARNING: Max OMP threads 1\n",
      "\n",
      "00:00 7.7Mb  CPU has 112 cores, running 1 threads\n",
      "00:00 7.7Mb   100.0% Calc posteriors\n",
      "00:01 34Mb    100.0% Consistency (1/2)\n",
      "00:01 34Mb    100.0% Consistency (2/2)\n",
      "00:01 34Mb    100.0% UPGMA5           \n",
      "00:01 34Mb    100.0% Refining\n",
      "\n",
      "muscle 5.1.linux64 []  1056Gb RAM, 112 cores\n",
      "Built May 13 2023 06:17:56\n",
      "(C) Copyright 2004-2021 Robert C. Edgar.\n",
      "https://drive5.com\n",
      "\n",
      "Input: 2 seqs, avg length 187, max 196\n",
      "\n",
      "\n",
      "WARNING: Max OMP threads 1\n",
      "\n",
      "00:00 7.7Mb  CPU has 112 cores, running 1 threads\n",
      "00:00 7.7Mb   100.0% Calc posteriors\n",
      "00:00 7.7Mb   100.0% UPGMA5         \n",
      "\n",
      "muscle 5.1.linux64 []  1056Gb RAM, 112 cores\n",
      "Built May 13 2023 06:17:56\n",
      "(C) Copyright 2004-2021 Robert C. Edgar.\n",
      "https://drive5.com\n",
      "\n",
      "Input: 2 seqs, avg length 520, max 752\n",
      "\n",
      "\n",
      "WARNING: Max OMP threads 1\n",
      "\n",
      "00:00 7.7Mb  CPU has 112 cores, running 1 threads\n",
      "00:00 7.7Mb   100.0% Calc posteriors\n",
      "00:00 7.7Mb   100.0% UPGMA5         \n",
      "\n",
      "muscle 5.1.linux64 []  1056Gb RAM, 112 cores\n",
      "Built May 13 2023 06:17:56\n",
      "(C) Copyright 2004-2021 Robert C. Edgar.\n",
      "https://drive5.com\n",
      "\n",
      "Input: 2 seqs, avg length 451, max 591\n",
      "\n",
      "\n",
      "WARNING: Max OMP threads 1\n",
      "\n",
      "00:00 7.7Mb  CPU has 112 cores, running 1 threads\n",
      "00:00 7.7Mb   100.0% Calc posteriors\n",
      "00:00 7.7Mb   100.0% UPGMA5         \n",
      "\n",
      "muscle 5.1.linux64 []  1056Gb RAM, 112 cores\n",
      "Built May 13 2023 06:17:56\n",
      "(C) Copyright 2004-2021 Robert C. Edgar.\n",
      "https://drive5.com\n",
      "\n",
      "Input: 2 seqs, avg length 925, max 1181\n",
      "\n",
      "\n",
      "WARNING: Max OMP threads 1\n",
      "\n",
      "00:00 7.7Mb  CPU has 112 cores, running 1 threads\n",
      "00:00 7.7Mb   100.0% Calc posteriors\n",
      "00:00 7.7Mb   100.0% UPGMA5         \n",
      "\n",
      "muscle 5.1.linux64 []  1056Gb RAM, 112 cores\n",
      "Built May 13 2023 06:17:56\n",
      "(C) Copyright 2004-2021 Robert C. Edgar.\n",
      "https://drive5.com\n",
      "\n",
      "Input: 2 seqs, avg length 1978, max 2163\n",
      "\n",
      "\n",
      "WARNING: Max OMP threads 1\n",
      "\n",
      "00:00 7.7Mb  CPU has 112 cores, running 1 threads\n",
      "00:00 7.7Mb   100.0% Calc posteriors\n",
      "00:00 7.7Mb   100.0% UPGMA5         \n",
      "\n",
      "muscle 5.1.linux64 []  1056Gb RAM, 112 cores\n",
      "Built May 13 2023 06:17:56\n",
      "(C) Copyright 2004-2021 Robert C. Edgar.\n",
      "https://drive5.com\n",
      "\n",
      "Input: 2 seqs, avg length 851, max 1378\n",
      "\n",
      "\n",
      "WARNING: Max OMP threads 1\n",
      "\n",
      "00:00 7.7Mb  CPU has 112 cores, running 1 threads\n",
      "00:00 7.7Mb   100.0% Calc posteriors\n",
      "00:00 7.7Mb   100.0% UPGMA5         \n",
      "\n",
      "muscle 5.1.linux64 []  1056Gb RAM, 112 cores\n",
      "Built May 13 2023 06:17:56\n",
      "(C) Copyright 2004-2021 Robert C. Edgar.\n",
      "https://drive5.com\n",
      "\n",
      "Input: 2 seqs, avg length 356, max 400\n",
      "\n",
      "\n",
      "WARNING: Max OMP threads 1\n",
      "\n",
      "00:00 7.7Mb  CPU has 112 cores, running 1 threads\n",
      "00:00 7.7Mb   100.0% Calc posteriors\n",
      "00:00 7.7Mb   100.0% UPGMA5         \n",
      "\n",
      "muscle 5.1.linux64 []  1056Gb RAM, 112 cores\n",
      "Built May 13 2023 06:17:56\n",
      "(C) Copyright 2004-2021 Robert C. Edgar.\n",
      "https://drive5.com\n",
      "\n",
      "Input: 2 seqs, avg length 354, max 395\n",
      "\n",
      "\n",
      "WARNING: Max OMP threads 1\n",
      "\n",
      "00:00 7.7Mb  CPU has 112 cores, running 1 threads\n",
      "00:00 7.7Mb   100.0% Calc posteriors\n",
      "00:00 7.7Mb   100.0% UPGMA5         \n",
      "\n",
      "muscle 5.1.linux64 []  1056Gb RAM, 112 cores\n",
      "Built May 13 2023 06:17:56\n",
      "(C) Copyright 2004-2021 Robert C. Edgar.\n",
      "https://drive5.com\n",
      "\n",
      "Input: 2 seqs, avg length 224, max 261\n",
      "\n",
      "\n",
      "WARNING: Max OMP threads 1\n",
      "\n",
      "00:00 7.7Mb  CPU has 112 cores, running 1 threads\n",
      "00:00 7.7Mb   100.0% Calc posteriors\n",
      "00:00 7.7Mb   100.0% UPGMA5         \n",
      "\n",
      "muscle 5.1.linux64 []  1056Gb RAM, 112 cores\n",
      "Built May 13 2023 06:17:56\n",
      "(C) Copyright 2004-2021 Robert C. Edgar.\n",
      "https://drive5.com\n",
      "\n",
      "Input: 2 seqs, avg length 836, max 1320\n",
      "\n",
      "\n",
      "WARNING: Max OMP threads 1\n",
      "\n",
      "00:00 7.7Mb  CPU has 112 cores, running 1 threads\n",
      "00:00 7.7Mb   100.0% Calc posteriors\n",
      "00:00 7.7Mb   100.0% UPGMA5         \n",
      "\n",
      "muscle 5.1.linux64 []  1056Gb RAM, 112 cores\n",
      "Built May 13 2023 06:17:56\n",
      "(C) Copyright 2004-2021 Robert C. Edgar.\n",
      "https://drive5.com\n",
      "\n",
      "Input: 2 seqs, avg length 347, max 392\n",
      "\n",
      "\n",
      "WARNING: Max OMP threads 1\n",
      "\n",
      "00:00 7.7Mb  CPU has 112 cores, running 1 threads\n",
      "00:00 7.7Mb   100.0% Calc posteriors\n",
      "00:00 7.7Mb   100.0% UPGMA5         \n",
      "\n",
      "muscle 5.1.linux64 []  1056Gb RAM, 112 cores\n",
      "Built May 13 2023 06:17:56\n",
      "(C) Copyright 2004-2021 Robert C. Edgar.\n",
      "https://drive5.com\n",
      "\n",
      "Input: 2 seqs, avg length 265, max 299\n",
      "\n",
      "\n",
      "WARNING: Max OMP threads 1\n",
      "\n",
      "00:00 7.7Mb  CPU has 112 cores, running 1 threads\n",
      "00:00 7.7Mb   100.0% Calc posteriors\n",
      "00:00 7.7Mb   100.0% UPGMA5         \n",
      "\n",
      "muscle 5.1.linux64 []  1056Gb RAM, 112 cores\n",
      "Built May 13 2023 06:17:56\n",
      "(C) Copyright 2004-2021 Robert C. Edgar.\n",
      "https://drive5.com\n",
      "\n",
      "Input: 2 seqs, avg length 236, max 249\n",
      "\n",
      "\n",
      "WARNING: Max OMP threads 1\n",
      "\n",
      "00:00 7.7Mb  CPU has 112 cores, running 1 threads\n",
      "00:00 7.7Mb   100.0% Calc posteriors\n",
      "00:00 7.7Mb   100.0% UPGMA5         \n",
      "\n",
      "muscle 5.1.linux64 []  1056Gb RAM, 112 cores\n",
      "Built May 13 2023 06:17:56\n",
      "(C) Copyright 2004-2021 Robert C. Edgar.\n",
      "https://drive5.com\n",
      "\n",
      "Input: 2 seqs, avg length 361, max 391\n",
      "\n",
      "\n",
      "WARNING: Max OMP threads 1\n",
      "\n",
      "00:00 7.7Mb  CPU has 112 cores, running 1 threads\n",
      "00:00 7.7Mb   100.0% Calc posteriors\n",
      "00:00 7.7Mb   100.0% UPGMA5         \n",
      "\n",
      "muscle 5.1.linux64 []  1056Gb RAM, 112 cores\n",
      "Built May 13 2023 06:17:56\n",
      "(C) Copyright 2004-2021 Robert C. Edgar.\n",
      "https://drive5.com\n",
      "\n",
      "Input: 2 seqs, avg length 1552, max 1644\n",
      "\n",
      "\n",
      "WARNING: Max OMP threads 1\n",
      "\n",
      "00:00 7.7Mb  CPU has 112 cores, running 1 threads\n",
      "00:00 7.7Mb   100.0% Calc posteriors\n",
      "00:00 7.7Mb   100.0% UPGMA5         \n",
      "\n",
      "muscle 5.1.linux64 []  1056Gb RAM, 112 cores\n",
      "Built May 13 2023 06:17:56\n",
      "(C) Copyright 2004-2021 Robert C. Edgar.\n",
      "https://drive5.com\n",
      "\n",
      "Input: 2 seqs, avg length 290, max 335\n",
      "\n",
      "\n",
      "WARNING: Max OMP threads 1\n",
      "\n",
      "00:00 7.7Mb  CPU has 112 cores, running 1 threads\n",
      "00:00 7.7Mb   100.0% Calc posteriors\n",
      "00:00 7.7Mb   100.0% UPGMA5         \n",
      "\n",
      "muscle 5.1.linux64 []  1056Gb RAM, 112 cores\n",
      "Built May 13 2023 06:17:56\n",
      "(C) Copyright 2004-2021 Robert C. Edgar.\n",
      "https://drive5.com\n",
      "\n",
      "Input: 2 seqs, avg length 440, max 510\n",
      "\n",
      "\n",
      "WARNING: Max OMP threads 1\n",
      "\n",
      "00:00 7.7Mb  CPU has 112 cores, running 1 threads\n",
      "00:00 7.7Mb   100.0% Calc posteriors\n",
      "00:00 7.7Mb   100.0% UPGMA5         \n",
      "\n",
      "muscle 5.1.linux64 []  1056Gb RAM, 112 cores\n",
      "Built May 13 2023 06:17:56\n",
      "(C) Copyright 2004-2021 Robert C. Edgar.\n",
      "https://drive5.com\n",
      "\n",
      "Input: 2 seqs, avg length 856, max 1377\n",
      "\n",
      "\n",
      "WARNING: Max OMP threads 1\n",
      "\n",
      "00:00 7.7Mb  CPU has 112 cores, running 1 threads\n",
      "00:00 7.7Mb   100.0% Calc posteriors\n",
      "00:01 7.7Mb   100.0% UPGMA5         \n",
      "\n",
      "muscle 5.1.linux64 []  1056Gb RAM, 112 cores\n",
      "Built May 13 2023 06:17:56\n",
      "(C) Copyright 2004-2021 Robert C. Edgar.\n",
      "https://drive5.com\n",
      "\n",
      "Input: 2 seqs, avg length 518, max 731\n",
      "\n",
      "\n",
      "WARNING: Max OMP threads 1\n",
      "\n",
      "00:00 7.7Mb  CPU has 112 cores, running 1 threads\n",
      "00:00 7.7Mb   100.0% Calc posteriors\n",
      "00:00 7.7Mb   100.0% UPGMA5         \n",
      "\n",
      "muscle 5.1.linux64 []  1056Gb RAM, 112 cores\n",
      "Built May 13 2023 06:17:56\n",
      "(C) Copyright 2004-2021 Robert C. Edgar.\n",
      "https://drive5.com\n",
      "\n",
      "Input: 2 seqs, avg length 481, max 651\n",
      "\n",
      "\n",
      "WARNING: Max OMP threads 1\n",
      "\n",
      "00:00 7.7Mb  CPU has 112 cores, running 1 threads\n",
      "00:00 7.7Mb   100.0% Calc posteriors\n",
      "00:00 7.7Mb   100.0% UPGMA5         \n",
      "\n",
      "muscle 5.1.linux64 []  1056Gb RAM, 112 cores\n",
      "Built May 13 2023 06:17:56\n",
      "(C) Copyright 2004-2021 Robert C. Edgar.\n",
      "https://drive5.com\n",
      "\n",
      "Input: 3 seqs, avg length 223, max 305\n",
      "\n",
      "\n",
      "WARNING: Max OMP threads 1\n",
      "\n",
      "00:00 7.7Mb  CPU has 112 cores, running 1 threads\n",
      "00:00 8.0Mb   100.0% Calc posteriors\n",
      "00:00 9.9Mb   100.0% Consistency (1/2)\n",
      "00:00 9.9Mb   100.0% Consistency (2/2)\n",
      "00:00 9.9Mb   100.0% UPGMA5           \n",
      "00:00 9.9Mb   100.0% Refining\n",
      "\n",
      "muscle 5.1.linux64 []  1056Gb RAM, 112 cores\n",
      "Built May 13 2023 06:17:56\n",
      "(C) Copyright 2004-2021 Robert C. Edgar.\n",
      "https://drive5.com\n",
      "\n",
      "Input: 2 seqs, avg length 513, max 799\n",
      "\n",
      "\n",
      "WARNING: Max OMP threads 1\n",
      "\n",
      "00:00 7.7Mb  CPU has 112 cores, running 1 threads\n",
      "00:00 7.7Mb   100.0% Calc posteriors\n",
      "00:00 7.7Mb   100.0% UPGMA5         \n",
      "\n",
      "muscle 5.1.linux64 []  1056Gb RAM, 112 cores\n",
      "Built May 13 2023 06:17:56\n",
      "(C) Copyright 2004-2021 Robert C. Edgar.\n",
      "https://drive5.com\n",
      "\n",
      "Input: 2 seqs, avg length 468, max 476\n",
      "\n",
      "\n",
      "WARNING: Max OMP threads 1\n",
      "\n",
      "00:00 7.7Mb  CPU has 112 cores, running 1 threads\n",
      "00:00 7.7Mb   100.0% Calc posteriors\n",
      "00:00 7.7Mb   100.0% UPGMA5         \n",
      "\n",
      "muscle 5.1.linux64 []  1056Gb RAM, 112 cores\n",
      "Built May 13 2023 06:17:56\n",
      "(C) Copyright 2004-2021 Robert C. Edgar.\n",
      "https://drive5.com\n",
      "\n",
      "Input: 2 seqs, avg length 515, max 652\n",
      "\n",
      "\n",
      "WARNING: Max OMP threads 1\n",
      "\n",
      "00:00 7.7Mb  CPU has 112 cores, running 1 threads\n",
      "00:00 7.7Mb   100.0% Calc posteriors\n",
      "00:00 7.7Mb   100.0% UPGMA5         \n",
      "\n",
      "muscle 5.1.linux64 []  1056Gb RAM, 112 cores\n",
      "Built May 13 2023 06:17:56\n",
      "(C) Copyright 2004-2021 Robert C. Edgar.\n",
      "https://drive5.com\n",
      "\n",
      "Input: 2 seqs, avg length 862, max 1335\n",
      "\n",
      "\n",
      "WARNING: Max OMP threads 1\n",
      "\n",
      "00:00 7.7Mb  CPU has 112 cores, running 1 threads\n",
      "00:00 7.7Mb   100.0% Calc posteriors\n",
      "00:00 7.7Mb   100.0% UPGMA5         \n",
      "\n",
      "muscle 5.1.linux64 []  1056Gb RAM, 112 cores\n",
      "Built May 13 2023 06:17:56\n",
      "(C) Copyright 2004-2021 Robert C. Edgar.\n",
      "https://drive5.com\n",
      "\n",
      "Input: 2 seqs, avg length 931, max 1181\n",
      "\n",
      "\n",
      "WARNING: Max OMP threads 1\n",
      "\n",
      "00:00 7.7Mb  CPU has 112 cores, running 1 threads\n",
      "00:00 7.7Mb   100.0% Calc posteriors\n",
      "00:00 7.7Mb   100.0% UPGMA5         \n",
      "\n",
      "muscle 5.1.linux64 []  1056Gb RAM, 112 cores\n",
      "Built May 13 2023 06:17:56\n",
      "(C) Copyright 2004-2021 Robert C. Edgar.\n",
      "https://drive5.com\n",
      "\n",
      "Input: 2 seqs, avg length 350, max 350\n",
      "\n",
      "\n",
      "WARNING: Max OMP threads 1\n",
      "\n",
      "00:00 7.7Mb  CPU has 112 cores, running 1 threads\n",
      "00:00 7.7Mb   100.0% Calc posteriors\n",
      "00:00 7.7Mb   100.0% UPGMA5         \n",
      "\n",
      "muscle 5.1.linux64 []  1056Gb RAM, 112 cores\n",
      "Built May 13 2023 06:17:56\n",
      "(C) Copyright 2004-2021 Robert C. Edgar.\n",
      "https://drive5.com\n",
      "\n",
      "Input: 2 seqs, avg length 750, max 1170\n",
      "\n",
      "\n",
      "WARNING: Max OMP threads 1\n",
      "\n",
      "00:00 7.7Mb  CPU has 112 cores, running 1 threads\n",
      "00:00 7.7Mb   100.0% Calc posteriors\n",
      "00:00 7.7Mb   100.0% UPGMA5         \n",
      "\n",
      "muscle 5.1.linux64 []  1056Gb RAM, 112 cores\n",
      "Built May 13 2023 06:17:56\n",
      "(C) Copyright 2004-2021 Robert C. Edgar.\n",
      "https://drive5.com\n",
      "\n",
      "Input: 4 seqs, avg length 386, max 646\n",
      "\n",
      "\n",
      "WARNING: Max OMP threads 1\n",
      "\n",
      "00:00 7.7Mb  CPU has 112 cores, running 1 threads\n",
      "00:00 8.9Mb   100.0% Calc posteriors\n",
      "00:00 9.0Mb   100.0% Consistency (1/2)\n",
      "00:00 9.0Mb   100.0% Consistency (2/2)\n",
      "00:00 9.0Mb   100.0% UPGMA5           \n",
      "00:00 10Mb    100.0% Refining\n",
      "\n",
      "muscle 5.1.linux64 []  1056Gb RAM, 112 cores\n",
      "Built May 13 2023 06:17:56\n",
      "(C) Copyright 2004-2021 Robert C. Edgar.\n",
      "https://drive5.com\n",
      "\n",
      "Input: 2 seqs, avg length 654, max 979\n",
      "\n",
      "\n",
      "WARNING: Max OMP threads 1\n",
      "\n",
      "00:00 7.7Mb  CPU has 112 cores, running 1 threads\n",
      "00:00 7.7Mb   100.0% Calc posteriors\n",
      "00:00 7.7Mb   100.0% UPGMA5         \n",
      "\n",
      "muscle 5.1.linux64 []  1056Gb RAM, 112 cores\n",
      "Built May 13 2023 06:17:56\n",
      "(C) Copyright 2004-2021 Robert C. Edgar.\n",
      "https://drive5.com\n",
      "\n",
      "Input: 2 seqs, avg length 344, max 427\n",
      "\n",
      "\n",
      "WARNING: Max OMP threads 1\n",
      "\n",
      "00:00 7.7Mb  CPU has 112 cores, running 1 threads\n",
      "00:00 7.7Mb   100.0% Calc posteriors\n",
      "00:00 7.7Mb   100.0% UPGMA5         \n"
     ]
    }
   ],
   "source": [
    "## run MUSCLE to obtain MSA\n",
    "for a in set(td['reference protein accession ']):\n",
    "    !sh pcr_align.sh \"$a\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "id": "c87c5f2e-8173-45bf-be9a-0f9bd80a976e",
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    },
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "mkdir: cannot create directory ‘PCR_result_alignments/XP_035708168.1’: File exists\n",
      "mkdir: cannot create directory ‘PCR_result_alignments/XP_046453153.1’: File exists\n",
      "mkdir: cannot create directory ‘PCR_result_alignments/XP_045027829.1’: File exists\n",
      "mkdir: cannot create directory ‘PCR_result_alignments/XP_046403459.1’: File exists\n",
      "mkdir: cannot create directory ‘PCR_result_alignments/XP_050513096.1’: File exists\n",
      "mkdir: cannot create directory ‘PCR_result_alignments/XP_042220148.1’: File exists\n",
      "mkdir: cannot create directory ‘PCR_result_alignments/XP_029735553.1’: File exists\n",
      "mkdir: cannot create directory ‘PCR_result_alignments/XP_046456339.1’: File exists\n",
      "mkdir: cannot create directory ‘PCR_result_alignments/XP_027232647.2’: File exists\n",
      "mkdir: cannot create directory ‘PCR_result_alignments/XP_023332299.1’: File exists\n",
      "mkdir: cannot create directory ‘PCR_result_alignments/XP_037026007.1’: File exists\n",
      "mkdir: cannot create directory ‘PCR_result_alignments/XP_023346081.1’: File exists\n",
      "mkdir: cannot create directory ‘PCR_result_alignments/XP_037041958.1’: File exists\n",
      "mkdir: cannot create directory ‘PCR_result_alignments/XP_037049533.1’: File exists\n",
      "mkdir: cannot create directory ‘PCR_result_alignments/XP_037030969.1’: File exists\n",
      "mkdir: cannot create directory ‘PCR_result_alignments/XP_035715507.1’: File exists\n",
      "mkdir: cannot create directory ‘PCR_result_alignments/XP_023328891.1’: File exists\n",
      "mkdir: cannot create directory ‘PCR_result_alignments/XP_046646423.1’: File exists\n",
      "mkdir: cannot create directory ‘PCR_result_alignments/XP_037051404.1’: File exists\n",
      "mkdir: cannot create directory ‘PCR_result_alignments/XP_011211954.1’: File exists\n",
      "mkdir: cannot create directory ‘PCR_result_alignments/XP_053634600.2’: File exists\n",
      "mkdir: cannot create directory ‘PCR_result_alignments/XP_032791327.2’: File exists\n",
      "mkdir: cannot create directory ‘PCR_result_alignments/XP_046445252.1’: File exists\n",
      "mkdir: cannot create directory ‘PCR_result_alignments/KAH9406650.1’: File exists\n",
      "mkdir: cannot create directory ‘PCR_result_alignments/XP_023324156.1’: File exists\n",
      "mkdir: cannot create directory ‘PCR_result_alignments/XP_001656415.1’: File exists\n",
      "mkdir: cannot create directory ‘PCR_result_alignments/XP_046649021.1’: File exists\n",
      "mkdir: cannot create directory ‘PCR_result_alignments/XP_046402901.1’: File exists\n",
      "mkdir: cannot create directory ‘PCR_result_alignments/XP_021699539.1’: File exists\n",
      "mkdir: cannot create directory ‘PCR_result_alignments/XP_037038002.1’: File exists\n",
      "mkdir: cannot create directory ‘PCR_result_alignments/XP_069990332.1’: File exists\n",
      "mkdir: cannot create directory ‘PCR_result_alignments/XP_023329593.1’: File exists\n",
      "mkdir: cannot create directory ‘PCR_result_alignments/XP_029821973.3’: File exists\n",
      "mkdir: cannot create directory ‘PCR_result_alignments/XP_015919340.2’: File exists\n"
     ]
    }
   ],
   "source": [
    "\n",
    "from pymsaviz import MsaViz\n",
    "import ast\n",
    "for ac in set(td['reference protein accession ']):\n",
    "    !mkdir 'PCR_result_alignments'/\"$ac\"\n",
    "    \n",
    "    mv = MsaViz(f\"PCR_result_alignments/{ac}/MSA_protein.fa\", wrap_length=100)\n",
    "    meta=[]\n",
    "    hgt=[]\n",
    "\n",
    "    for inter in ast.literal_eval(df_all[df_all.protein==ac].Metazoan_intervals.values[0]):\n",
    "        meta.append(inter)\n",
    "    for inter in ast.literal_eval(df_all[df_all.protein==ac].HGT_intervals.values[0]):\n",
    "        hgt.append(inter)\n",
    "    for m in meta:\n",
    "        mv.add_text_annotation((int(m[0]), int(m[1])), \"Metazoan\", text_color=\"orange\", range_color=\"orange\")\n",
    "    for h in hgt:\n",
    "        mv.add_text_annotation((int(h[0]), int(h[1])), \"HGT\", text_color=\"purple\", range_color=\"purple\")\n",
    "    mv.savefig(f\"PCR_result_alignments/{ac}/MSA_protein.png\", dpi=400)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "id": "58b0beee-2f69-4982-bba8-b06508ce529d",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "PDF generated successfully: PCR_result_alignments_merged.pdf\n"
     ]
    }
   ],
   "source": [
    "from fpdf import FPDF\n",
    "from PIL import Image\n",
    "import os\n",
    "\n",
    "# Path to the directory containing SVG files (here used for PNG images)\n",
    "input_directory = \"PCR_result_alignments\"\n",
    "\n",
    "# Output PDF file\n",
    "pdf_output = \"PCR_result_alignments_merged.pdf\"\n",
    "\n",
    "# Create a PDF document\n",
    "pdf = FPDF()\n",
    "pdf.set_auto_page_break(auto=True, margin=15)\n",
    "\n",
    "# For demonstration, we'll process one image.\n",
    "# Replace `set(mapdf.reference_protein)` with your actual iterable if needed.\n",
    "for a in set(td['reference protein accession ']):\n",
    "    pdf.add_page()\n",
    "    pdf.set_font(\"Arial\", size=11)\n",
    "    pdf.cell(200, 10, txt=a, ln=True, align='C')\n",
    "    \n",
    "    # Construct the full path to the image file\n",
    "    image_path = os.path.join(input_directory, a, \"MSA_protein.png\")\n",
    "    \n",
    "    # Open the image using Pillow to get its dimensions (in pixels)\n",
    "    with Image.open(image_path) as img:\n",
    "        img_width, img_height = img.size\n",
    "\n",
    "    # Convert pixels to millimeters.\n",
    "    # Assuming an image DPI of 96. (1 inch = 25.4 mm, so conversion factor = 25.4 / DPI.)\n",
    "    dpi = 96  # Modify this if your images have a different DPI setting.\n",
    "    conv_factor = 25.4 / dpi\n",
    "    img_width_mm = img_width * conv_factor\n",
    "    img_height_mm = img_height * conv_factor\n",
    "\n",
    "    # Define available width and height for the image within the PDF page.\n",
    "    # This calculation accounts for the margins. You can modify these values as needed.\n",
    "    max_width = pdf.w - 2 * pdf.l_margin\n",
    "    # Use the top margin offset (pdf.t_margin + 20) for header space, then subtract the bottom margin.\n",
    "    max_height = pdf.h - (pdf.t_margin + 20) - pdf.b_margin\n",
    "\n",
    "    # Calculate the scaling factor to fit the image within the available space,\n",
    "    # keeping the original aspect ratio.\n",
    "    scale = min(max_width / img_width_mm, max_height / img_height_mm, 1)\n",
    "    new_width = img_width_mm * scale\n",
    "    new_height = img_height_mm * scale\n",
    "\n",
    "    # Place the image onto the page with the new dimensions.\n",
    "    pdf.image(image_path, x=pdf.l_margin, y=pdf.t_margin + 20, w=new_width, h=new_height)\n",
    "    \n",
    "   \n",
    "\n",
    "# Save the PDF to a file.\n",
    "pdf.output(pdf_output)\n",
    "print(f\"PDF generated successfully: {pdf_output}\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0eb055ab-b59e-4dbc-b58a-c0e7c320b234",
   "metadata": {},
   "outputs": [],
   "source": [
    "input_file = open(f\"genbank_translations_may_2025.fa\")\n",
    "fasta_dict = SeqIO.to_dict(SeqIO.parse(input_file, \"fasta\"))\n",
    "fasta_dict = {x:fasta_dict[x] for x in td['sequence name ']}\n",
    "records = [fasta_dict[i] for i in fasta_dict]\n",
    "with open(\"genbank_submission_07_23_2025/genbank_translations.fa\", \"w\") as out:\n",
    "    SeqIO.write(records, out, \"fasta\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3e95b796-800d-485c-a8f4-56cd53b05bae",
   "metadata": {},
   "outputs": [],
   "source": [
    "input_file = open(f\"genbank_translations_may_2025.fa\")\n",
    "fasta_dict = SeqIO.to_dict(SeqIO.parse(input_file, \"fasta\"))\n",
    "fasta_dict = {x:fasta_dict[x] for x in td['sequence name ']}\n",
    "records = [fasta_dict[i] for i in fasta_dict]\n",
    "with open(\"genbank_submission_07_23_2025/genbank_translations.fa\", \"w\") as out:\n",
    "    SeqIO.write(records, out, \"fasta\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4a13dc4d-d607-4602-90a8-47c32f1dbafe",
   "metadata": {},
   "source": [
    "## add genbank accessions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0ef28a56-efb1-44bd-bdaf-5a22689edc69",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "79a2fa21-d9f9-44f9-84ac-b37c6086c92c",
   "metadata": {},
   "outputs": [],
   "source": [
    "td=pd.read_csv(\"SI tables/sequencing_results.tsv\",sep=\"\\t\",index_col=0)\n",
    "gb=pd.read_csv(\"genbank_submission_07_23_2025/genbank_submission_07_23_2025_genbank_submit.tsv\",sep=\"\\t\",index_col=0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b145417f-3f68-4409-8e30-822ebc58622c",
   "metadata": {},
   "outputs": [],
   "source": [
    "##load mapping between submitted sequence accession and genbank id, provided in email from genbank\n",
    "gen=\"\"\"\n",
    "SUB15265677 Dapuli_chimera2_seq1        PV989571\n",
    "SUB15265677 Brcopr_chimera1_seq1        PV989572\n",
    "SUB15265677 Dapuli_chimera3_seq1        PV989573\n",
    "SUB15265677 Typutr_chimera1_seq1        PV989574\n",
    "SUB15265677 Typutr_chimera1_seq2        PV989575\n",
    "SUB15265677 Damagn_chimera1_seq1        PV989576\n",
    "SUB15265677 Damagn_chimera1_seq2        PV989577\n",
    "SUB15265677 Iseleg_chimera1_seq1        PV989578\n",
    "SUB15265677 Chquad_chimera1_seq1        PV989579\n",
    "SUB15265677 Divirg_chimera1_seq1        PV989580\n",
    "SUB15265677 Dapule_chimera1_seq1        PV989581\n",
    "SUB15265677 Dapule_chimera2_seq1        PV989582\n",
    "SUB15265677 Dapule_chimera3_seq1        PV989583\n",
    "SUB15265677 Iseleg_chimera2_seq1        PV989584\n",
    "SUB15265677 Patepi_chimera1_seq1        PV989585\n",
    "SUB15265677 Hoamer_chimera1_seq1        PV989586\n",
    "SUB15265677 Brcopr_chimera2_seq1        PV989587\n",
    "SUB15265677 Brcopr_chimera3_seq1        PV989588\n",
    "SUB15265677 Brcopr_chimera4_seq1        PV989589\n",
    "SUB15265677 Brcopr_chimera5_seq1        PV989590\n",
    "SUB15265677 Brcopr_chimera6_seq1        PV989591\n",
    "SUB15265677 Focand_chimera2_seq1        PV989592\n",
    "SUB15265677 Focand_chimera4_seq1        PV989593\n",
    "SUB15265677 Focand_chimera4_seq2        PV989594\n",
    "SUB15265677 Focand_chimera4_seq3        PV989595\n",
    "SUB15265677 Damagn_chimera2_seq1        PV989596\n",
    "SUB15265677 Ixscap_chimera1_seq1        PV989597\n",
    "SUB15265677 Aealbo_chimera1_seq1        PV989598\n",
    "SUB15265677 Pevann_chimera1_seq1        PV989599\n",
    "SUB15265677 Pevann_chimera2_seq1        PV989600\n",
    "SUB15265677 Eucaro_chimera1_seq1        PV989601\n",
    "SUB15265677 Eucaro_chimera2_seq1        PV989602\n",
    "SUB15265677 Eucaro_chimera3_seq1        PV989603\n",
    "SUB15265677 Eucaro_chimera3_seq2        PV989604\n",
    "SUB15265677 Eucaro_chimera3_seq3        PV989605\n",
    "SUB15265677 Eucaro_chimera4_seq1        PV989606\n",
    "SUB15265677 Eucaro_chimera5_seq1        PV989607\n",
    "SUB15265677 Aeaegy_chimera1_seq1        PV989608\n",
    "SUB15265677 Badors_chimera2_seq1        PV989609\n",
    "SUB15265677 Aeaegy_chimera2_seq1        PV989610\"\"\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "28eaf793-2b28-498c-be03-387b70fbac12",
   "metadata": {},
   "outputs": [],
   "source": [
    "genmap={x.split(\"        \")[0].split(\" \")[1]:x.split(\"        \")[1].strip() for x in gen.split(\"\\n\")[1:]}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9ecd9bbc-82d7-4606-bda6-e00e589f2dd3",
   "metadata": {},
   "outputs": [],
   "source": [
    "for index, row in td.iterrows():\n",
    "    if index in genmap:\n",
    "        td.loc[index,'genbank accession']=genmap[index]\n",
    "td.to_csv(\"SI tables/sequencing_results.tsv\",sep=\"\\t\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "fffe5c4d-1121-4068-a671-03f5d44e9df6",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "s13=pd.read_csv(\"SI tables/SI_table_13_sequencing_results.tsv\",sep=\"\\t\",index_col=0)\n",
    "for index, row in s13.iterrows():\n",
    "    s13.loc[index,'chimera cluster #']=df_all[df_all.protein==row['reference protein accession']]['cluster'].values[0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "419b684c-7d1e-4f2c-be16-4ff084245526",
   "metadata": {},
   "outputs": [],
   "source": [
    "s13.to_csv(\"SI tables/SI_table_13_sequencing_results.tsv\",sep=\"\\t\")"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python [conda env:.conda-rishabh]",
   "language": "python",
   "name": "conda-env-.conda-rishabh-py"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
