{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "fe89e93f-6b5f-46cc-953f-824cd18349ec",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pickle\n",
    "import pandas as pd\n",
    "import multiprocessing as mp\n",
    "import numpy as np\n",
    "import pickle\n",
    "import matplotlib.pyplot as plt\n",
    "from Bio import SeqIO\n",
    "import os\n",
    "import subprocess\n",
    "import ast\n",
    "import warnings\n",
    "from pathlib import Path\n",
    "warnings.filterwarnings('ignore')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "aa4c6f38-65fc-4c5e-9df7-bee7ea8bc3f8",
   "metadata": {},
   "outputs": [],
   "source": [
    "## define mapping between taxanomy and color of leaf on tree\n",
    "import matplotlib.pyplot as plt\n",
    "import matplotlib.colors as mcolors\n",
    "colors=[\"blue\",\"brown\",\"green\",\"pink\",\"orange\",\"black\",\"red\",\"yellow\",\"magenta\",'grey']\n",
    "cmap={c:mcolors.to_hex(c) for c in colors}\n",
    "color_to_taxonomy_legend={\n",
    "    \"blue\": \"Bacteria\",\n",
    "    \"brown\": \"Fungi\",\n",
    "    \"green\":\"Viridiplantae\",\n",
    "    \"yellow\" : \"Viruses\",\n",
    "    \"orange\": \"Non-Arthropod Metazoa\",\n",
    "    \"pink\": \"Arthropod (non-chimera)\",\n",
    "    \"magenta\" : \"Arthropod chimera (primary)\",\n",
    "    \"red\": \"Arthropod chimera (secondary)\",\n",
    "    \"black\": \"Other\",\n",
    "    'grey': 'Rotifera'\n",
    "    \n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "id": "fb5bc96f-29e3-488e-bef3-eca181063d9e",
   "metadata": {},
   "outputs": [],
   "source": [
    "\"\"\"\n",
    "takes a row of a dataframe with a 'lineage' and 'secondary' column (latter for secondary chimera, boolean)\n",
    "returns a color label\n",
    "\"\"\"\n",
    "def color(row):\n",
    "    x=row['lineage']\n",
    "    if 'Fungi' in x:\n",
    "        return 'brown'\n",
    "    elif 'Viruses' in x:\n",
    "        return 'yellow'\n",
    "    elif 'Bacteria' in x:\n",
    "        return 'blue'\n",
    "    elif 'Viridiplantae' in x:\n",
    "        return 'green'\n",
    "    elif 'primary_chimera' in str(row):\n",
    "        return 'magenta'\n",
    "    elif 'secondary' in row.index and row['secondary']=='True':\n",
    "        return 'red'\n",
    "    elif 'Arthropoda' in x:\n",
    "        return 'pink'\n",
    "    elif 'Rotifera' in x:\n",
    "        return 'grey'\n",
    "    elif 'Metazoa' in x:\n",
    "        return 'orange'\n",
    "    else:\n",
    "        return 'black'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "32823982-5ea5-4368-b94f-04c182f97ada",
   "metadata": {},
   "outputs": [],
   "source": [
    "\"\"\"\n",
    "adds a legend to the itol_color_strip.txt annotation file\n",
    "\"\"\"\n",
    "def add_legend(interv,dftax):\n",
    "    ch=\";\".join(interv.split(\";\")[0:2])\n",
    "    file_path=f\"outputs/phylogenetic_dataset/{ch}/{interv}/itol_color_strip.txt\"\n",
    "    f=open(file_path,\"r\").readlines()\n",
    "    if \"LEGEND_TITLE Taxonomic labels\\n\" not in f:\n",
    "        \n",
    "        colorsi=list(set(dftax['color']))\n",
    "\n",
    "        cmapi={cmap[x]:color_to_taxonomy_legend[x] for x in colorsi}\n",
    "        cmapi = dict(sorted(cmapi.items(), key=lambda item: item[1]))\n",
    "        legend_text=\"LEGEND_TITLE Taxonomic labels\\n\"\n",
    "        shapes=\"LEGEND_SHAPES \"\n",
    "        for i in range(len(cmapi)):\n",
    "            shapes+=(\"1 \")\n",
    "        shapes=shapes[:-1]\n",
    "        legend_text=legend_text+\"\\n\"+shapes\n",
    "        legend_colors=\"LEGEND_COLORS \"\n",
    "        for c in cmapi:\n",
    "            legend_colors+=c+\" \"\n",
    "        legend_colors=legend_colors[:-1]\n",
    "        legend_text=legend_text+\"\\n\"+legend_colors\n",
    "        legend_LABELS=\"LEGEND_LABELS \"\n",
    "        for c in cmapi:\n",
    "            legend_LABELS+=cmapi[c].replace(\" \",\"_\")+\" \"\n",
    "        legend_LABELS=legend_LABELS[:-1]\n",
    "        legend_text=legend_text+\"\\n\"+legend_LABELS+\"\\n\"+\"\\n\"\n",
    "        legend_text=legend_text+\"STRIP_WIDTH 40\\n\\n\"\n",
    "        with open(file_path, \"r\") as file:\n",
    "            lines = file.readlines()\n",
    "\n",
    "        for i, line in enumerate(lines):\n",
    "            if line.strip() == \"DATA\":\n",
    "                insert_index = i\n",
    "                break\n",
    "        lines.insert(insert_index, legend_text)\n",
    "        with open(file_path, \"w\") as file:\n",
    "            file.writelines(lines)\n",
    "    else:\n",
    "        print(\"present\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "ea67587d-f962-497d-8317-16bd9181ae60",
   "metadata": {},
   "outputs": [],
   "source": [
    "\"\"\"\n",
    "applies minimum ancestor deviation rooting using a script from https://www.nature.com/articles/s41559-017-0193\n",
    "\"\"\"\n",
    "def root_tree(interv):\n",
    "    ch=\";\".join(interv.split(\";\")[0:2])\n",
    "    subprocess.run([\"python3\", \"scripts/mad.py\", f'outputs/phylogenetic_dataset/{ch}/{interv}/edited_ml_tree.treefile' ,\"-n\"]) \n",
    "    tbs=Tree(f'outputs/phylogenetic_dataset/{ch}/{interv}/edited_ml_tree.treefile')\n",
    "    trs=Tree(f'outputs/phylogenetic_dataset/{ch}/{interv}/edited_ml_tree.treefile.rooted')\n",
    "\n",
    "    ## use MAD-rooted tree to outgroup root tree with bootstraps (MAD.py removes )\n",
    "    for n in trs.iter_descendants(\"preorder\"):\n",
    "        l =[ str(leaf) for leaf in n]\n",
    "        l2=[x.split(\"--\")[1] for x in l]\n",
    "        break\n",
    "    if len(l2)>1:\n",
    "        tbs.set_outgroup(tbs.get_common_ancestor(l2))\n",
    "    else:\n",
    "        tbs.set_outgroup(l2[0])\n",
    "    \n",
    "    newick_str = tbs.write(format=0)  \n",
    "    # optional: save to file\n",
    "    with open(f'outputs/phylogenetic_dataset/{ch}/{interv}/final_rooted.tree', \"w\") as fh:\n",
    "        fh.write(newick_str)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "b8465ffa-1f54-4a36-ae05-fa714fe41488",
   "metadata": {},
   "outputs": [],
   "source": [
    "import sys\n",
    "import itolapi\n",
    "from itolapi import Itol\n",
    "\"\"\"\n",
    "Upload tree and associated annotation files to itol\n",
    "\"\"\"\n",
    "def upload_to_itol(interv):\n",
    "    ch=\";\".join(interv.split(\";\")[0:2])\n",
    "    current_dir = Path(os.path.realpath(f\"outputs/phylogenetic_dataset/{ch}/{interv}/final_rooted.tree\")).parent\n",
    "    tree = current_dir/\"final_rooted.tree\"\n",
    "\n",
    "\n",
    "    root_path = current_dir.parent\n",
    "    sys.path.append(str(root_path))\n",
    "    test = Itol()\n",
    "\n",
    "    # Set the tree file\n",
    "\n",
    "\n",
    "    test.add_file(current_dir / 'itol_color_strip.txt')\n",
    "    test.add_file(current_dir / 'itol_taxonomic_info.txt')\n",
    "    test.add_file(tree)\n",
    "    # Add parameters\n",
    "\n",
    "    test.params['APIkey'] = 'INSERT_API_KEY'\n",
    "    test.params['treeName'] = interv\n",
    "    if 'HGT' in interv:\n",
    "        test.params['projectName'] =\"HGT_trees\"\n",
    "    else:\n",
    "        test.params['projectName'] =\"Metazoan_trees\"\n",
    "    test.params['workspaceName'] = \"Arthropod_chimeric_HGT_ML_trees\"\n",
    "\n",
    "    # Submit the tree, return false if fail\n",
    "    return test.upload()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "id": "0277f807-dfb7-4c6c-b556-4ce7ca6d428f",
   "metadata": {},
   "outputs": [],
   "source": [
    "\"\"\"\n",
    "Master function to annotate, root, and upload trees\n",
    "\"\"\"\n",
    "def upload_main(interv):\n",
    "    ch=\";\".join(interv.split(\";\")[0:2])\n",
    "    \n",
    "    ##Load taxonomic info dataframe \n",
    "    combined=pd.read_csv(f'outputs/phylogenetic_dataset/{ch}/{interv}/combined_sequences_data.tsv',sep='\\t')\n",
    "    try:\n",
    "        combined=combined.set_index('target_name')\n",
    "    except:\n",
    "        combined=combined.set_index('sseqid')\n",
    "\n",
    "\n",
    "    tree = Tree(f'outputs/phylogenetic_dataset/{ch}/{interv}/ml_tree.treefile', format=0)\n",
    "    leaf_names = tree.get_leaf_names()\n",
    "    for index, row in combined.iterrows():\n",
    "        try:\n",
    "            combined.loc[index, 'itol_name']=[x for x in leaf_names if index.replace(\";\",\"_\") in x][0]\n",
    "        except:\n",
    "            continue\n",
    "\n",
    "\n",
    "    #correct tree leaf names (itol automatically removes characters after \",\"\n",
    "\n",
    "    new_name_map={i:\n",
    "    x.split(';')[1] if ';' in x else x\n",
    "    for i,x in zip(combined.itol_name, combined.index) }    \n",
    "\n",
    "\n",
    "    combined=combined[combined.lineage.astype(str)!='None']\n",
    "    \n",
    "    ##add color based on taxonomic info\n",
    "    for index, row in combined.iterrows():\n",
    "        combined.loc[index, 'color']=color(row)\n",
    "\n",
    "    for leaf in tree.iter_leaves():\n",
    "         leaf.name = new_name_map[leaf.name]\n",
    "\n",
    "\n",
    "    newick_str = tree.write(format=0)  \n",
    "\n",
    "    with open(f'outputs/phylogenetic_dataset/{ch}/{interv}/edited_ml_tree.treefile', \"w\") as fh:\n",
    "        fh.write(newick_str)\n",
    "    \n",
    "    ## write itol annotation file for colors\n",
    "    f=open(f\"outputs/phylogenetic_dataset/{ch}/{interv}/itol_color_strip.txt\",\"w\")\n",
    "    f1=open(\"scripts/itol_color_strip_template.txt\",\"r\")\n",
    "    for x in f1.readlines():\n",
    "        f.write(x)\n",
    "    f1.close()\n",
    "    for index, row in combined.iterrows():\n",
    "        f.write(new_name_map[row['itol_name']]+\" \"+cmap[row.color])\n",
    "        f.write(\"\\n\")\n",
    "    f.close()\n",
    " \n",
    "    ## write itol annotation file for taxonomic annotation\n",
    "    f=open(f\"outputs/phylogenetic_dataset/{ch}/{interv}/itol_taxonomic_info.txt\",\"w\")\n",
    "    f1=open(\"scripts/itol_taxonomic_info_template.txt\",\"r\")\n",
    "    for x in f1.readlines():\n",
    "        f.write(x)\n",
    "    f1.close()\n",
    "    for index, row in combined.iterrows():\n",
    "        s=row['species'].replace(\" \",\"_\")\n",
    "        a=f\"{s};{row['order']};{row['class']};{row['phylum']}\"\n",
    "        a=a.replace(',','_')\n",
    "        f.write(new_name_map[row['itol_name']]+\",\"+a+\",-1,#000000,normal,1,0\")\n",
    "        f.write(\"\\n\")\n",
    "    f.close()\n",
    "    \n",
    "    ## call MAD rooting tree\n",
    "    root_tree(interv)\n",
    "    \n",
    "    ## add legend for colors \n",
    "    add_legend(interv,combined)\n",
    "    \n",
    "    ##uplaod\n",
    "    upload_to_itol(interv)\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "13fec4de-9af1-4a62-b3da-429f5014d80d",
   "metadata": {},
   "outputs": [],
   "source": [
    "from pathlib import Path\n",
    "\n",
    "def find_parents_with_contree(root):\n",
    "    \"\"\"Return a sorted list of parent directories whose child contains\n",
    "    an 'ml_tree.contree' file.\n",
    "    \"\"\"\n",
    "    root = Path(root)            # <-- coerce str â†’ Path\n",
    "    parents: set[Path] = set()\n",
    "\n",
    "    for child_dir in root.rglob(\"*\"):\n",
    "        if child_dir.is_dir() and (child_dir / \"ml_tree.contree\").is_file():\n",
    "            parents.add(str(child_dir.resolve()).split(\"/\")[-1])\n",
    "\n",
    "    return sorted(str(p) for p in parents)\n",
    "completed_intervals = find_parents_with_contree(\"outputs/phylogenetic_dataset\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "b9a5d925-4ebc-4746-ba9e-69a632c766c0",
   "metadata": {},
   "outputs": [],
   "source": [
    "##Upload HGT intervals\n",
    "td=[x for x in completed_intervals if 'HGT' in x]\n",
    "with mp.Pool(40) as pool:\n",
    "    pool.map(upload_main,td)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "4cf3b4a9-f80d-4ab7-907a-bb5ca520936d",
   "metadata": {},
   "outputs": [],
   "source": [
    "##Upload HGT intervals\n",
    "completed_intervals = find_parents_with_contree(\"outputs/phylogenetic_dataset\")\n",
    "td=[x for x in completed_intervals if 'Meta' in x]\n",
    "with mp.Pool(40) as pool:\n",
    "    pool.map(upload_main,td)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "242669e7-c36b-4319-af71-cabade27d62e",
   "metadata": {},
   "source": [
    "## Collect relevant data for dryad upload"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "b1436d3a-b784-450f-937b-dd78830256ab",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "hgt=list(pd.read_csv(\"Tree_manual_inspection_HGT.tsv\",sep='\\t',index_col=0).index)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "c28872b1-a6cf-480b-a6a4-2168921933a1",
   "metadata": {},
   "outputs": [],
   "source": [
    "!mkdir 'outputs/phylogenetic_data_filtered'\n",
    "for x in hgt:\n",
    "    d= f'outputs/phylogenetic_data_filtered/{x}'\n",
    "    !mkdir -p \"$d\"\n",
    "    ch=\";\".join(x.split(\";\")[0:2])\n",
    "    od=f\"outputs/phylogenetic_dataset/{ch}/{x}\"\n",
    "    !cp \"$od\"/\"final_rooted.tree\" \"$d\"/\"final_rooted.tree\"\n",
    "    !cp \"$od\"/\"ml_tree.iqtree\" \"$d\"/\"ml_tree.iqtree\"\n",
    "    !cp \"$od\"/\"itol_color_strip.txt\" \"$d\"/\"itol_color_strip.txt\"\n",
    "    !cp \"$od\"/\"itol_taxonomic_info.txt\" \"$d\"/\"itol_taxonomic_info.txt\"\n",
    "    !cp \"$od\"/\"trimmed_MSA.fasta\" \"$d\"/\"trimmed_MSA.fasta\"\n",
    "    !cp \"$od\"/\"MSA.fasta\" \"$d\"/\"MSA.fasta\"\n",
    "    !cp \"$od\"/\"all_sequences.fa\" \"$d\"/\"all_sequences.fa\"\n",
    "    !cp \"$od\"/\"combined_sequences_data.tsv\" \"$d\"/\"combined_sequences_data.tsv\"\n",
    "   "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "789de97a-423c-46b7-a88a-a44d70f07f47",
   "metadata": {},
   "outputs": [],
   "source": [
    "meta=list(pd.read_csv(\"Tree_manual_inspection_Metazoan.tsv\",sep='\\t',index_col=0).index)\n",
    "for x in meta:\n",
    "    d= f'outputs/phylogenetic_data_filtered/{x}'\n",
    "    !mkdir -p \"$d\"\n",
    "    ch=\";\".join(x.split(\";\")[0:2])\n",
    "    od=f\"outputs/phylogenetic_dataset/{ch}/{x}\"\n",
    "    !cp \"$od\"/\"final_rooted.tree\" \"$d\"/\"final_rooted.tree\"\n",
    "    !cp \"$od\"/\"ml_tree.iqtree\" \"$d\"/\"ml_tree.iqtree\"\n",
    "    !cp \"$od\"/\"itol_color_strip.txt\" \"$d\"/\"itol_color_strip.txt\"\n",
    "    !cp \"$od\"/\"itol_taxonomic_info.txt\" \"$d\"/\"itol_taxonomic_info.txt\"\n",
    "    !cp \"$od\"/\"trimmed_MSA.fasta\" \"$d\"/\"trimmed_MSA.fasta\"\n",
    "    !cp \"$od\"/\"MSA.fasta\" \"$d\"/\"MSA.fasta\"\n",
    "    !cp \"$od\"/\"all_sequences.fa\" \"$d\"/\"all_sequences.fa\"\n",
    "    !cp \"$od\"/\"combined_sequences_data.tsv\" \"$d\"/\"combined_sequences_data.tsv\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "38cda4ce-734b-458a-ba68-995fb861d084",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash\n",
    "tar -czvf phylogenetic_data_filtered.tar.gz \\\n",
    "    -C outputs \\\n",
    "    phylogenetic_data_filtered"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "69ed388c-971a-4868-8eef-f8092de3dc61",
   "metadata": {},
   "source": [
    "## Tree upload for manuscript 8/27/2025\n",
    "correct mis-labeling of some secondary chimeras on the tree, correct title of legend, upload all trees to one itol project"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "10ad7932-215c-4e55-a487-730efbd90db6",
   "metadata": {},
   "outputs": [],
   "source": [
    "si2=pd.read_csv(\"SI tables/SI table 2.tsv\",sep=\"\\t\",index_col=0)\n",
    "##extract header\n",
    "intervals=[]\n",
    "import ast\n",
    "for index, row in si2[si2.representative].iterrows():\n",
    "    for h in ast.literal_eval(row.HGT_intervals):\n",
    "        intervals.append(index+\";HGT_\"+str(h).replace(\" \",\"\"))\n",
    "    for h in ast.literal_eval(row.Metazoan_intervals):\n",
    "        intervals.append(index+\";Meta_\"+str(h).replace(\" \",\"\"))\n",
    "        \n",
    "    \n",
    "import os\n",
    "l=[x for x in os.listdir(\"outputs/phylogenetic_data_filtered\") if '.ipynb' not in x]\n",
    "\n",
    "for xi in l:\n",
    "    df=pd.read_csv(f\"outputs/phylogenetic_data_filtered/{xi}/combined_sequences_data.tsv\",sep=\"\\t\")\n",
    "   \n",
    "    try:\n",
    "\n",
    "        for index, row in df[df.target_name.isin((sec.index))].iterrows():\n",
    "            df.loc[index,'secondary']=True\n",
    "    except:\n",
    "      \n",
    "        for index, row in df[df.sseqid.isin((sec.index))].iterrows():\n",
    "            df.loc[index,'secondary']=True\n",
    "    df.to_csv(f\"outputs/phylogenetic_data_filtered/{xi}/combined_sequences_data.tsv\",sep=\"\\t\")   \n",
    "##remove unnecessary columns from header\n",
    "for xi in l:\n",
    "    df=pd.read_csv(f\"outputs/phylogenetic_data_filtered/{xi}/combined_sequences_data.tsv\",sep=\"\\t\")\n",
    "    df=df.drop([x for x in df.columns if 'Unnamed' in x],axis=1)\n",
    "    df.to_csv(f\"outputs/phylogenetic_data_filtered/{xi}/combined_sequences_data.tsv\",sep=\"\\t\")  \n",
    "\n",
    "for xi in l:\n",
    "    df=pd.read_csv(f\"outputs/phylogenetic_data_filtered/{xi}/combined_sequences_data.tsv\",sep=\"\\t\",index_col=0)\n",
    "    c=list(df.columns)\n",
    "    c[0]='index'\n",
    "    df.columns=c\n",
    "    df=df.set_index('index')\n",
    "    df.to_csv(f\"outputs/phylogenetic_data_filtered/{xi}/combined_sequences_data.tsv\",sep=\"\\t\")  \n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "28d939f6-59e0-4b90-98b6-ae2d5a0da1e9",
   "metadata": {},
   "outputs": [],
   "source": [
    "## correct itol labels\n",
    "\n",
    "for interv in l:\n",
    "    df=pd.read_csv(f\"outputs/phylogenetic_data_filtered/{interv}/combined_sequences_data.tsv\",sep=\"\\t\")\n",
    "    file_path=f\"outputs/phylogenetic_data_filtered/{interv}/itol_color_strip.txt\"\n",
    "    f=open(file_path,\"r\").readlines()\n",
    "    f2=open(file_path,\"w\")\n",
    "    for x in f:\n",
    "        if 'Taxonomic labels' in x:\n",
    "            x=x.replace('Taxonomic labels','Taxonomic_labels')\n",
    "        f2.write(x)\n",
    "    f2.close()\n",
    "    if 'secondary' in df.columns:\n",
    "        if 'True' in set(df.secondary.astype(str)):\n",
    "            try:\n",
    "                secs=set([x.split(\";\")[1] for x in df[df.secondary.astype(str)=='True']['target_name']])\n",
    "            except:\n",
    "                secs=set([x.split(\";\")[1] for x in df[df.secondary.astype(str)=='True']['sseqid']])\n",
    "            f2=open(file_path,\"w\")\n",
    "            for x in f:\n",
    "                if 'LEGEND_COLORS' in x and \"#ff0000\" not in x:\n",
    "                    x=x.strip()+\" #ff0000\\n\"\n",
    "                if 'LEGEND_LABELS' in x and 'Arthropod_chimera_(secondary)' not in x:\n",
    "                    x=x.strip()+\" Arthropod_chimera_(secondary)\\n\"\n",
    "                elif x.split(\" \")[0] in set(secs):\n",
    "                    x=x.split(\" \")[0]+\" #ff0000\\n\"\n",
    "\n",
    "                f2.write(x)\n",
    "            f2.close()\n",
    "import sys\n",
    "import itolapi\n",
    "from itolapi import Itol\n",
    "from pathlib import Path\n",
    "\"\"\"\n",
    "Upload tree and associated annotation files to itol\n",
    "\"\"\"\n",
    "def upload_to_itol(interv):\n",
    "\n",
    "    ch=\";\".join(interv.split(\";\")[0:2])\n",
    "    current_dir = Path(os.path.realpath(f\"outputs/phylogenetic_data_filtered/{interv}/final_rooted.tree\")).parent\n",
    "    tree = current_dir/\"final_rooted.tree\"\n",
    "\n",
    "\n",
    "    root_path = current_dir.parent\n",
    "    sys.path.append(str(root_path))\n",
    "    test = Itol()\n",
    "\n",
    "    # Set the tree file\n",
    "\n",
    "\n",
    "    test.add_file(current_dir / 'itol_color_strip.txt')\n",
    "    test.add_file(current_dir / 'itol_taxonomic_info.txt')\n",
    "    test.add_file(tree)\n",
    "    # Add parameters\n",
    "\n",
    "    test.params['APIkey'] = 'INSERT KEY'\n",
    "    if interv in intervals:\n",
    "        interv=\"*\"+interv\n",
    "    test.params['treeName'] = interv\n",
    "    test.params['projectName'] =\"Arthropod HGT-chimera interval trees 8/27/2025\"\n",
    "    test.params['workspaceName'] = \"Arthropod_chimeric_HGT_ML_trees\"\n",
    "\n",
    "    # Submit the tree, return false if fail\n",
    "    return test.upload()\n",
    "import multiprocessing as mp\n",
    "with mp.Pool(40) as pool:\n",
    "    pool.map(upload_to_itol,l)\n",
    "        \n",
    "        "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "cd465d30-1388-4789-b4d2-d206510076b0",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash\n",
    "tar -czvf phylogenetic_data_filtered.tar.gz \\\n",
    "    -C outputs \\\n",
    "    phylogenetic_data_filtered"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python [conda env:.conda-rishabh]",
   "language": "python",
   "name": "conda-env-.conda-rishabh-py"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
